mpack-instance-manager/src/main/python/instance_manager/instance_manager.py

```diff
@@ -0,0 +1,500 @@
+  result = {}
+  for instance_name in instances_dict:
+    result[instance_name] = instances_dict[instance_name].build_json_output(output_conf_dir, output_path)
+
+  return {plural_name: result}
+
+
+# FIXME use mpack.json to determine category
+def find_module_category(path, module_name):
+  walk = os.walk(os.path.join(path, module_name))
+  if CONFIGS_DIRECTORY_NAME in next(walk)[1]:
+    return "CLIENT"
+  return "SERVER"
+
+
+def build_json_output(instances, output_conf_dir=False, output_path=False):
+  result = {}
+  for mpack_name in instances.keys():
+    result[mpack_name] = build_json_output_from_instances_dict(instances[mpack_name], MpackInstance.plural_name,
+                                                               output_conf_dir, output_path)
+  return {'mpacks': result}
+
+
+class MetaMpack:
+  def __init__(self, name, version, component_module_map, mpack_json):
+    self.name = name
+    self.version = version
+    self.component_module_map = component_module_map
+    self.mpack_json = mpack_json
+    self.module_categoty_map = self.parse_mpack_json_into_module_category_map()
+    self.module_component_types_map = self.parse_mpack_json_into_module_component_types_map()
+
+  def parse_mpack_json_into_module_category_map(self):
+    result = {}
+    if not self.mpack_json:
+      return None
+    for module in self.mpack_json['modules']:
+      result[module['id']] = module['category']
+
+    return result
+
+  def parse_mpack_json_into_module_component_types_map(self):
+    result = {}
+    if not self.mpack_json:
+      return None
+    for module in self.mpack_json['modules']:
+      result[module['id']] = []
+      for component in module['components']:
+        result[module['id']].append(component['id'])
+
+    return result
+
+  def get_component_category(self, component_type):
+    for module in self.module_component_types_map:
+      if component_type in self.module_component_types_map[module]:
+        return self.module_categoty_map[module]
+    return None
+
+  @staticmethod
+  def parse_mpack(path, name, version):
+    # build components map from soft links
+    component_module_map = {}
+    for filename in os.listdir(path):
+      if os.path.islink(os.path.join(path, filename)):
+        component_module_map[filename] = os.path.realpath(os.path.join(path, filename))
+
+    return MetaMpack(name=name,
+                     version=version,
+                     component_module_map=component_module_map,
+                     mpack_json=json.load(open(os.path.join(path, MPACK_JSON_FILE_NAME))))
+
+  @staticmethod
+  def parse_mpacks(path):
+    result = {}
+    walk = os.walk(path)
+    for mpack_name in next(walk)[1]:
+      result[mpack_name] = MetaMpack.parse_into_mpack_objects(os.path.join(path, mpack_name))
+    return result
+
+  @staticmethod
+  def parse_into_mpack_objects(path):
+    result = {}
+    walk = os.walk(path)
```

Comment:
os.walk doesnt not work with unicode paths. Check AMBARI-22819. However most likely it's okay for us. I don't think customers will have unicode paths in mpack forlders.